Added cmake tooling

pthom opened the pull request:
If you are interested, this is a CMake file that I added. This CMake file was tested under linux, macOS and windows. 
I also added a backend "glfw_gl3" for samples.

sammyfreg commented on Code/ServerApp/Source/NetImguiServer_App.h:
```diff
@@ -6,9 +6,30 @@
-#define HAL_API_PLATFORM_GLFW_GL3		0														// Currently only compiles in release (library include compatibility)
-#define HAL_API_PLATFORM_SOKOL			0
+// 1. Either edit the lines below to select the desired backend
+#if !defined(NETIMGUI_SERVER_APP_BACKEND_SOKOL) && !defined(NETIMGUI_SERVER_APP_BACKEND_GLFW_GL3) && !defined(NETIMGUI_SERVER_APP_BACKEND_WIN32_DX11)
```
I think that instead of having to list every case here, it should be a series of ifdef / else until the default behaviour is kept in the final else, without having to rely on defines.

```
#if defined(xxx)
// xxx
#elif defined(yyy)
// yyy
#elif ....
#else
// Default values here
#endif
```

pthom replied:
Ok, the code now reads 

```cpp
// Select the backend from with cmake (with -DNETIMGUI_SERVER_APP_BACKEND_WIN32_DX11, etc)
// Note : HAL_API_PLATFORM_GLFW_GL3 currently only compiles in release (library include compatibility)
#if defined(NETIMGUI_SERVER_APP_BACKEND_WIN32_DX11)
    #define HAL_API_PLATFORM_WIN32_DX11	1
    #define HAL_API_PLATFORM_SOKOL		0
    #define HAL_API_PLATFORM_GLFW_GL3	0
#elif defined(NETIMGUI_SERVER_APP_BACKEND_GLFW_GL3)
    #define HAL_API_PLATFORM_WIN32_DX11	0
    #define HAL_API_PLATFORM_SOKOL		0
    #define HAL_API_PLATFORM_GLFW_GL3	1
#elif(defined(NETIMGUI_SERVER_APP_BACKEND_SOKOL))
    #define HAL_API_PLATFORM_WIN32_DX11	0
    #define HAL_API_PLATFORM_SOKOL		1
    #define HAL_API_PLATFORM_GLFW_GL3	0
#else
    // Or manually edit the lines below to select the desired backend
    #define HAL_API_PLATFORM_WIN32_DX11		1
    #define HAL_API_PLATFORM_GLFW_GL3		0
    #define HAL_API_PLATFORM_SOKOL			0
#endif
```

sammyfreg commented on CMakeLists.txt:
```diff
@@ -0,0 +1,213 @@
+cmake_minimum_required(VERSION 3.15)
```
I'm thinking about wanting this CMake file included or not. On one hand, it's good work and definitely nice to have, but on the other hand it might soon get out of date without me dedicating extra time to maintain it too. It is not too complex (the Sharpmake version is a bit more complicated, but could also be used to output a makefile I believe).

sammyfreg replied:
The other thing is that we really only need the NetImguiServer application to have multi-platform compiling support. The samples are just a simple demonstration that can easily be compiled without a build system configured, and the NetImgui client code is usually included in an existing project, and is the one that was meant to be truly multi-platform.

sammyfreg commented on Code/Sample/Common/main_glfw_gl3.cpp:
```diff
@@ -0,0 +1,335 @@
+// Dear ImGui: standalone example application for GLFW + OpenGL 3, using programmable pipeline
```
All files in the hierarchy are added to the build. This file doesn't compile under Windows without OpenGL lib setup `Cannot open include file: 'GLFW/glfw3.h'`. The cpp file content needs to be wrapped in a define so that its content is only active with a OPENGL define (and probably something similar for the default DirectX samples).

pthom replied:
Ok,  I added a #ifdef on top of main_glfw_gl3.cpp and main.cpp

pthom commented:
@sammyfreg wrote:
> I'm thinking about wanting this CMake file included or not. On one hand, it's good work and definitely nice to have, but on the other hand it might soon get out of date without me dedicating extra time to maintain it too. It is not too complex (the Sharpmake version is a bit more complicated, but could also be used to output a makefile I believe).

My advice is that you should not merge this PR, and leave it opened so that people who are looking for it might find it and know that they need to adapt it because it might be outdated.

pthom commented:
@sammyfreg wrote:
> The other thing is that we really only need the NetImguiServer application to have multi-platform compiling support. The samples are just a simple demonstration that can easily be compiled without a build system configured, and the NetImgui client code is usually included in an existing project, and is the one that was meant to be truly multi-platform.

By default, here are the options in the cmake file:

```cmake
option(NETIMGUI_BUILD_CLIENT "Build client" ON)
option(NETIMGUI_BUILD_IMGUI "Build imgui from sources in netImgui" ON)
option(NETIMGUI_BUILD_SAMPLES "Build samples" ${PROJECT_IS_TOP_LEVEL})
option(NETIMGUI_BUILD_SERVER_LIB "Build server lib" ${PROJECT_IS_TOP_LEVEL})
option(NETIMGUI_BUILD_SERVER_APP_SOKOL "Build server app" ${PROJECT_IS_TOP_LEVEL})
```

i.e. 
- the client lib is built by default
- ImGui is also built by default, but users can set `-DNETIMGUI_BUILD_IMGUI=OFF` to use their own
- Sample are built only if this is not included as a sub-project
- The server is built only if building netImgui as the main project. In the current state, the server can only be built with Sokol, but users may adapt it quite easily: I left hints in the cmake file for that.

pthom commented:
I think I had been too fast in the last push, I added an ifdef inside `Code/Sample/Common/main.cpp` which probably led the CI to fail. I removed it and I hope that the CI will pass. 

Anyhow, it is not very important that it passes since this PR is meant to remain opened and not merged.

starlight-traveler commented:
Real easy to add glfw3 support, tests on MacOS with glfw3 installed via brew (not tested on non-Darwin Unix, will update once I get around to that). DPI is a bit wonky, but that can be fixed in menu.

Will leave this here for posterity/those who dislike Cmake files. Simply replace the function not implemented.

> function(_add_net_imgui_server_app_glfw_gl3)
>     set(server_app_glfwgl3_dir ${SERVER_DIR}/GlfwGL3)
>     file(GLOB server_app_glfwgl3_files ${server_app_glfwgl3_dir}/*.cpp)
>     add_executable(net_imgui_server_app_glfwgl3 ${server_app_glfwgl3_files})
>     target_link_libraries(net_imgui_server_app_glfwgl3 net_imgui_server_lib)
> 
>     # Find the GLFW package
>     find_package(glfw3 REQUIRED)
> 
>     if(APPLE)
>         # Link GLFW and frameworks needed for macOS
>         target_link_libraries(net_imgui_server_app_glfwgl3 glfw "${GLFW_LIBRARIES}")
>     elseif(UNIX)
>         # Link GLFW and libraries needed for Unix-like systems
>         target_link_libraries(net_imgui_server_app_glfwgl3 glfw "${GLFW_LIBRARIES}")
>         target_link_libraries(net_imgui_server_app_glfwgl3 X11 GL Xcursor Xi)
>     endif()
> 
> endfunction()

sammyfreg commented:
What do you mean with 'wonki DPI'? It might be that you need to implement DPI awareness on a non window platform. 

Search for **@SAMPLE_EDIT (DPI Awareness)** in the source code (the @SAMPLE_EDIT tags indicate where I made some small changes in the original **Dear ImGui's** sample code) to see how it is handled. Basically, the platform code should be calling **NetImguiServer::UI::SetWindowDPI** to let the Server Application what's the current Window DPI, to adjust the font size.
